File: src/main/java/learnyouakotlin/part4/SignupSheet.java
package learnyouakotlin.part4;

import org.jetbrains.annotations.NotNull;

import java.util.LinkedHashSet;
import java.util.Set;

public class SignupSheet {
    private SessionId sessionId;
    private int capacity;
    private final LinkedHashSet<AttendeeId> signups = new LinkedHashSet<>();
    private boolean isClosed = false;

    public SignupSheet() {
    }
    
    public SignupSheet(@NotNull SessionId sessionId, int capacity) {
        this.sessionId = sessionId;
        this.capacity = capacity;
    }

    public SessionId getSessionId() {
        return sessionId;
    }

    public void setSessionId(SessionId sessionId) {
        if (sessionId != null) {
            throw new IllegalStateException("you cannot change the sessionId after it has been set");
        }
        this.sessionId = sessionId;
    }

    public int getCapacity() {
        return capacity;
    }

    public void setCapacity(int newCapacity) {
        if (capacity != 0) {
            throw new IllegalStateException("you cannot change the capacity after it has been set");
        }

        this.capacity = newCapacity;
    }

    public boolean isFull() {
        return signups.size() == capacity;
    }

    public boolean isClosed() {
        return isClosed;
    }

    public Set<AttendeeId> getSignups() {
        return Set.copyOf(signups);
    }

    public boolean isSignedUp(AttendeeId attendeeId) {
        return signups.contains(attendeeId);
    }

    public void signUp(AttendeeId attendeeId) {
        if (isClosed()) throw new IllegalStateException("sign-up has closed");
        if (isFull()) throw new IllegalStateException("session is full");

        signups.add(attendeeId);
    }

    public void cancelSignUp(AttendeeId attendeeId) {
        if (isClosed()) throw new IllegalStateException("sign-up has closed");

        signups.remove(attendeeId);
    }

    public void close() {
        isClosed = true;
    }
}

File: src/main/java/learnyouakotlin/part4/SignupSheet.kt
package learnyouakotlin.part4

class SignupSheet {
    var capacity = 0
        set(value) {
            check(capacity == 0) { "you cannot change the capacity after it has been set" }
            field = value
        }

    var sessionId: SessionId? = null
        set(value) {
            check(sessionId == null) { "you cannot change the sessionId after it has been set" }
            field = value
        }

    var signups = emptySet<AttendeeId>()
        private set

    var isClosed = false
        private set

    constructor()
    constructor(sessionId: SessionId, capacity: Int) {
        this.sessionId = sessionId
        this.capacity = capacity
    }

    private val isFull: Boolean
        get() = signups.size == capacity

    fun close() {
        isClosed = true
    }

    fun isSignedUp(attendeeId: AttendeeId): Boolean = attendeeId in signups

    fun signUp(attendeeId: AttendeeId) {
        check(!isClosed) { "sign-up has closed" }
        check(!isFull) { "session is full" }
        signups = signups + attendeeId
    }

    fun cancelSignUp(attendeeId: AttendeeId) {
        check(!isClosed) { "sign-up has closed" }
        signups = signups - attendeeId
    }
}
